### DISCUSSION TITLE
Nightscout on Google Cloud using Docker

### DISCUSSION THREAD

**Discussion by bjornoleh** [NightscoutFoundation/xDrip#2460] [Ideas]
_Edited to add:_
___________________
 **Warning:**
This is **not an official guide, and there is no support provided.** Please understand that you are “on your own” if you do apply it. 

This is just a **proof of concept** of an alternative approach to the xDrip guide for Nightscout on Google Cloud. Do not ask the xDrip team for support if you get stuck when diverting from their documentation.

This guide **will** cause conflicts if you have already followed the xDrip guide. 

You may not be able to follow the xDrip guide on the same VM after following my instructions.
 You may have to delete the VM and start over. Please backup and download to your computer any database contents you wish to keep before deleting the VM.

And finally: Please don’t share this guide. I do not want people to start relying on this guide without understanding the above, and then go and bug the xDrip team about it if they get stuck.

________________

Hi,

I just wanted to let you know I have put together a guide for how to set up NS on Google Cloud using Docker. It is heavily based on your xDrip NS instructions, with some modifications. 

This method has some benefits over the deployment used in the xDrip script. In my experience, it is much faster to deploy and update, seems to require less machine resources, and can be used to host multiple NS instances with very little additional effort. The official Nightscout docker image is used, with the "latest" tag. This is updated automatically when the NS team push a new release to the master branch.  To update your Nightscout, just SSH into your VM, and type `docker compose up -d`.

 It is also possible to specify a fixed version by changing the tag. This is defined in docker-compose.yml, and a change will take action after  `docker compose up -d`.

Both the Nightscout app and the mongo database are running in Docker containers, and this seems to also have some benefits. The database can "easily" and also securely be accessed from tools like MongoDB Compass.

I believe it will still fit the Compute Engine free tier, but this is yet to be verified, as I have run it for less than a week. Looks promising, though. 

I don’t really want to take on an additional support burden, and ideally, this could be incorporated in the xDrip Nightscout setup guide if you find it useful, so I don't have to maintain yet another set of instructions for users to choose from. But I'll share it for early testing here and probably also on Nightscout Discord etc.

The guide is in this Google document (available upon request):
[Nightscout on Google Cloud using Docker](https://docs.google.com/document/d/16V29ZlJbzRHmI-avVrA_Jo1czWKgoY2oEOnf9OfVVOY/edit?usp=sharing&authuser=0)

There are lots of links to the source, so I am not trying to take any credit for that. I'll put in some more explicit references for the screenshots if you like. This is only the first draft anyways. 

Thanks for your guides, by the way, I would not have managed this without. I am happy with my setup, and have as such reached my goal. But wanted to share this in case it can be helpful for others too, as I did spend quite a bit of time to make it work like I wanted.

**Comment by Navid200** [comment]
I wish you had not done this.  

We can add Docker to our setup if we want to.
When you create a copy, there will be people who see it and  think it's the original.  We are still working on our setup.  We are not done yet.

There are people who are looking for the easiest solution to set up and be done with.
Some are going to follow your guide and run into trouble, like any other approach.  What will happen to them?  Will you be able to support them?
And what would be the reason?
How does this benefit them?

If your objective was to state that our setup can also be used with a Docker, you could have stated just that.  You didn't have to create a guide.

If you believe that we are wrong and there is a significant advantage in using Docker, you could still state just that keeping our guide as a reference and only stating that using a docker is significantly better and focus your contribution to the community in explaining that.

**Reply by bjornoleh** [reply]
There is absolutely no harm done, this is the reason I have not  distributed this anywhere except here. 

And I have now restricted the access to the Google document. If anyone wants access, they can ping me and get it by promising not to share it and that they have understood they are on their own if using it, and not bother the xDrip team about it. If people get in trouble, they can always delete the VM and start over.

**Reply by Navid200** [reply]
Thanks

**Reply by bjornoleh** [reply]
And to explain again very briefly what issues my suggestion is trying to solve:

1. This method can very easily be used to set up multiple NS instances on the same VM. This is very relevant, as many families have multiple members with t1d. This could be solved in many ways, of course, but possibly a Docker based solution will be more compatible with the limitations of the Google Compute Engine free tier limits (I don't know if this is true, just my speculation)..

2. The Docker deployment is very fast, and no significant downtime is required for Nightscout updates. The roughly 30 minutes downtime required for Nightscout updates following the current xDrip guide may be acceptable in most cases, but can be inconvenient enough at times. If an update should fail and it has to be repeated, the downtime would be significant.

**Comment by tzachi-dar** [comment]
Does one have to prepare the docker before actually deploying it?
Can the deployment be done using menus only.
The entire nightscout on google work is done in order to solve one issue:
There are some people who need to add a line to a file, and they can not do it.
By the way, I have the filling they are 99% of the population.
With that in mind, is the docket solution viable. For example, "running two dockers on one VM is possible, you just need to update the nginx.conf file (very easy with sed)" is not possible.

**Reply by bjornoleh** [reply]
Absolutely, the deploy of Nightscout from a Docker image can be done from a menu without any manual interaction, just like the current xDrip guide. 

The whole setup is defined in [docker-compose.yml](https://github.com/bjornoleh/ns-setup/blob/multi/docker-compose.yml). This only need editing if NS config variables are to be changed.

In the case of the setup in my guide, the[ add.sh](https://github.com/bjornoleh/ns-setup/blob/multi/add.sh) script is appending settings for individual NS instances to docker-compose.yml, together with traefik configurations for each subdomain

The deployment is simply `sudo docker compose up -d`, and can be performed from the main script or the menu for updating.

The official Nightscout docker image is launched, and the Nightscout instance is updated if there has been an update to the image. In this script I use the "latest" tag. A specific version can be used too if being extra conservative and fearing a broken master release.

https://hub.docker.com/r/nightscout/cgm-remote-monitor/tags

My script (which is forked from the [ns-setup repo of justmara](https://github.com/justmara/ns-setup) and just slightly adapted) only require three inputs from the user:

1. email address for (SSL?) certificates, managed by traefik
2. domain name
3. subdomain nalme

I expect users to need an explanation of subdomain vs domain. But apart from that ir should be easy enough. Editing config variables will be the crux for both my approach and the current xDrip guide (although the heroku vars script will help with this).

**Reply by Navid200** [reply]
If Docker is all advantage, why does the alternative exist?

**Reply by bjornoleh** [reply]
Legacy? I don't expect older methods to disappear just because something new emerges. And some will be more comfortable with the old methods. I really cant answer if there are downsides to Docker, but I have not found any so far (after very brief testing!).

The Nightscout on Azure guide is also based on docker, by the way.

**Reply by Navid200** [reply]
So, you mean Sulka knows that Docker is better in every respect and he continues to release Nightscout as he does because ...?

**Reply by bjornoleh** [reply]
I think you are confusing things here. The code base is the same. The NS app can then be built with different methods, such as npm or Docker. The Nightscout Docker image linked above is automatically built and uploaded to Docker hub at every release, and at every dev commit for the dev tag. It is the NS team who owns the account on Docker hub. There is no conflict here!

**Reply by bjornoleh** [reply]
And it was Sulka who set it up…

https://github.com/nightscout/cgm-remote-monitor/commits/master/.github/workflows/main.yml

**Reply by bjornoleh** [reply]
https://github.com/nightscout/cgm-remote-monitor/blob/master/.github/workflows/main.yml

**Reply by tzachi-dar** [reply]
Is the docker running on the vm that is running on the cloud? (That is does it have an overhead ?)

**Reply by bjornoleh** [reply]
The docker container is running on the VM. My understanding is that this is well optimised and has minimal overhead. Nightscout, MongoDB and traefik are all running in the same container. Seems very robust.

**Reply by Navid200** [reply]
Can a user create a fork of the Nightscout on docker and customize the plugins before deploying it in their virtual machine as they can with our current setup?

**Reply by bjornoleh** [reply]
Yes, more or less. I do this. You’ll have to build the docker image and push it to your own Docker hub repo. You can build from the command line after installing stuff, or build in the GitHub cloud using GitHub actions, just like the Nightscout team does. This is achievable for those who actually customise their Nightscout.

**Reply by bjornoleh** [reply]
So the user would select their Docker repo/image name and tag instead of GitHub repo and branch.

**Comment by bjornoleh** [comment]
I notice I am exceeding the free tier somehow. At least according to the report. But its less than 1 Euro or USD in one week, so really not a big deal. 

I have four live NS sites on the same VM, two of them receive looping data from FreeAPS X (FAX), and I know there is an issue with FAX doing excessive uploads/syncing of treatments. Some of this may also be related to mongodump/mongoimport of 400 Mb data, and some for messing around in Linux, installing some packages etc.

![image](https://user-images.githubusercontent.com/63544115/202295697-c925d13c-3ec0-4919-8bff-4a6cf9c413c4.png)

**Reply by Navid200** [reply]
Would you please post a screenshot of the status page?

**Reply by bjornoleh** [reply]
Which status page is that? I can look again tomorrow.

**Reply by Navid200** [reply]
Item number 9:
https://navid200.github.io/xDrip/docs/Nightscout/Menu.html

**Reply by bjornoleh** [reply]
Ok, thanks. I'm not using the menu system, but could run the Status.sh script by

`curl https://raw.githubusercontent.com/jamorham/nightscout-vps/vps-1/Status.sh | bash`

**Reply by bjornoleh** [reply]
![image](https://user-images.githubusercontent.com/63544115/202390447-4a798fe3-ac7d-4c8d-acda-9f68d773077d.png)

**Reply by bjornoleh** [reply]
I'm pretty sure it is set up correctly for the free tier.

**Reply by Navid200** [reply]
What does it show if you click on this?
![202295697-c925d13c-3ec0-4919-8bff-4a6cf9c413c4](https://user-images.githubusercontent.com/51497406/202473043-db5afb0e-ad3b-42f2-97f7-6e0ad1c47be3.png)

**Reply by bjornoleh** [reply]
It does not show anything interesting in that viev:

Current month
1–17 November 2022
Month-to-date total cost
NOK6.71
End-of-month total cost (forecasted)
NOK21.11

**Reply by bjornoleh** [reply]
The costs per day are really low (1 NOK = 0,1 USD) , and so far today I made money on this (negative cost). Not sure how accurate it is, and what the bill will be in the end of month. Not too worried, but interesting to learn more. I have not been able to find a breakdown which show what part of my usage that is exceeding the free tier limits.

![image](https://user-images.githubusercontent.com/63544115/202486429-1e7fac1c-8ab4-4574-b6bb-f000573db2f7.png)

**Reply by Navid200** [reply]
Well, I asked you to post a screenshot, which you didn't.  

I see that you have 2 projects.  Why is that?

**Reply by bjornoleh** [reply]
Okay then, here is the screenshot. Please point me to the interesting bits.

![image](https://user-images.githubusercontent.com/63544115/202497059-debb70a7-975b-4f03-96a7-d820e1b0c835.png)


I had a project within the free tier a couple of years ago. That instance has been shut down since, and I deleted the project before starting this one. I can probably still see the historical data for it, but there is nothing new going on there.

**Reply by Navid200** [reply]
> I had a project within the free tier a couple of years ago. That instance has been shut down since, 

If you go to that project, is there an instance regardless of it running or not? 

Was there cost in that project when you ran it?

**Reply by bjornoleh** [reply]
There were never any costs on that project. And I have only one project and one VM instance now. The old one is deleted, the VM no longer exists.

**Reply by Navid200** [reply]
Is the machine in one of the three specified regions that qualify for free tier?

**Reply by bjornoleh** [reply]
Yes, us-east1

**Comment by tzachi-dar** [comment]
This page explains how to see what is costing the money:
https://cloud.google.com/billing/docs/how-to/reports
In any case, if you will choose the cost table and cost breakdown, they
should show much more information on where the money is being spent.

**Reply by bjornoleh** [reply]
Thanks! I found how to get more details. By default, the chart is grouped by project, so no detail really. If grouping by SKU, one can see what the cost was for. In my case, it was network egress from the data centre to EMEA. 

![272BAEF8-45EE-4238-A1F7-786B990168CE](https://user-images.githubusercontent.com/63544115/202566654-b5fb1e41-a21c-42de-85ee-51e747ce43c9.jpeg)

**Reply by bjornoleh** [reply]
EMEA seems like the right region (Europe, Middle East and Africa). Europe in my case.

**Comment by bjornoleh** [comment]
A small group of testers (only seven so far) have requested access to the guide. 

I am curious if someone have set up multiple NS sites using this guide, and if the setup appears to stay within the free tier when multiple APS systems (Loop, AAPS etc) are uploading data. 

For myself, I seem to be exceeding the free tier slightly, but this may be due to some issue with the APS system we are using (FreeAPS X (FAX)). We have two phones running FAX and uploading to individual NS instances, and another two NS instances which are fed by Dexcom Share data. 

I have a forecasted monthly cost of around 3 USD (or EUR).

For comparison, the E2 instance core + ram we are getting for free with the free tier is worth around 6 USD per month. 

You can see this by grouping costs by SKU. 

![ED7E3DD5-45F2-4E14-8295-3DAB3C3410AA](https://user-images.githubusercontent.com/63544115/203374118-475ac38c-680c-465d-830c-e3259206b6f4.jpeg)

**Reply by Navid200** [reply]
I don't believe there is any violations of the free tier limit caused by uploading too much.
The limit is on egress, which means outgoing traffic from your Google Cloud virtual machine.

**Reply by bjornoleh** [reply]
Yes, that’s true. I don’t know, but suspect FreeAPS X is generating outbound traffic when syncing treatments? The other alternative is that a follower app like Loop Follow is generating excessive egress. Will be interesting to see if others have the same or different experiences.

**Reply by Navid200** [reply]
But, I don't think people had more than one Nightscout using one Heroku account.  So, why is this a concern?  If someone has multiple diabetic kids, they can have multiple Google accounts, no?

**Reply by bjornoleh** [reply]
Don’t you agree it’s easier to maintain one VM and have all config variables in a single file? That’s what I have now. I don’t really care if it will cost me 3 USD per month, but as I said I suspect it could even be free for other setups.

**Reply by bjornoleh** [reply]
Also, it was easy to set up multiple Heroku accounts and manage them from a single account. Then one could deploy updates and set up config variables for each instance from this one account.

**Comment by marionbarker** [comment]
Two question/comments:

What is the advantage of docker over 3 separate VM? I think Bjorn said it was faster to update to new NS version with that approach. The other would be not having to go through the VM setup task again. 

I am not showing a charge expected for 3 separate VM. 

Details:

I spun up 3 VM on google cloud. One is used by my real phone. Two are connected to two test phones each running Loop-dev, each uploading to their unique subdomain 24/7. I have two raspberry pi in the house so using. DASH simulator for those 2 test phones. One test phone uploads CGM simulator data to its site. The other test phone connects to my Dexcom Share via bridge. Both test phones are running closed loop. 

To further separate the testing, each is in a separate project. Two of the project imported > 1 GB from my t1pal site. They are each showing about $0.05. The one project where I did not import data is showing $0.00.

**Reply by bjornoleh** [reply]
My understanding is that you can run a single Google Cloud Compute Engine free tier VM continuously for the full month, or several VMs for a total run time of one month. If this is true, you should expect to incur costs after ten days with three VM instance. Not sure if you have let them run for this long yet?

Docker excels in ease of setup, speed of deployment and updates, and presumably it could run more efficiently on the available machine resources. Not sure if the last point is true though. 

My docker-compose.yml file contains all configurations including database structure and NS config variables. I could deploy this on another VM very easily without any additional configuration as long as I have a backup of this file.

**Reply by Navid200** [reply]
You can have multiple Google accounts.  In each, you can have a free virtual machine.

**Reply by bjornoleh** [reply]
Absolutely, but that’s not very practical. Would you be able to easily switch between accounts using the Google Cloud app on a phone?

**Reply by Navid200** [reply]
It's perfectly practical.
Most people never look at Nightscout after it is up and running.

I have been using Nightscout for years.  I never looked at Heroku until a fer months ago when it was announced that it will not be free anymore.

I see no reason to worry about this.

**Reply by bjornoleh** [reply]
I notice it’s possible to add additional accounts to the Google Cloud app, that definitely helps

**Comment by marionbarker** [comment]
@bjornoleh - I have not run any of my VM for even 10 days yet.
@Navid200 - I am running 3 projects under one account, each with one VM.

**Reply by bjornoleh** [reply]
That’s possible, but it won’t be free. I think it’s around 6 USD per instance if you were to pay for it. I don’t know how the cost would be calculated though.

**Reply by bjornoleh** [reply]
If you test my guide, it’s best to start with a fresh VM. Otherwise you’ll get some conflicts.

**Comment by bjornoleh** [comment]
Hi, just a heads up if anyone is actually using a Nightscout setup based on my Google document posted above:

In general, this has been working very well for me. Except that after a month or two, the disc of my VM filled up to the point of no longer being able to boot. I repeated the setup, and saw that it was about to happen again, and it also happened to two others that I am aware of.

Turns out this was due to a lack of any limits to logs retention in the docker-compose.yml template that my setup and was based on, (from the Nightscout/cgm-remote-monitor repo). Given enough time, any disc will eventually fill up without any limits. 

My setup appears to generate more logs than most (FreeAPS X is known to do that), and I was hopefully bitten by this before too many others had to deal with it.

To avoid any such issues, there are two options:

1. Periodically type the following in your VM terminal, which will purge the logs:

`sudo docker compose down && sudo docker compose up -d`

2. For a proper fix, define limits for log retention in your docker-compose.yml, similarly to this PR:

https://github.com/nightscout/cgm-remote-monitor/pull/7884/commits/4dd6f3960a3243a109e061e727cd1e3d822ea6f6

You will need this in the top of the docker-compose:

```
x-logging:
  &default-logging
  options:
    max-size: '10m'
    max-file: '5'
  driver: json-file
```

And then append this to each service (mongo, traefik and each instance of Nightscout):

`logging: *default-logging`

The new logging limits only take effect after the containers are recreated:

`sudo docker compose down && sudo docker compose up -d`

The chosen parameters for log limits are a bit arbitrary, but can easily be changed if needed: 
```
    max-size: '10m'
    max-file: '5'
```

I have updated my branch that is used in the Google document, so it should work for anyone starting from scratch. I didn't test this though, so please let me know of any issues.

This is still not meant to be the guide of all guides for NS on Docker (another one is in the making by someone else), but this information is kind of pertinent, so wanted to get that out to anyone who might be bitten by this issue soon (or have already been bitten...).

**Comment by bjornoleh** [comment]
Hi, another heads up if anyone is actually using a Nightscout setup based on my Google document posted above:

Please consider removing the port mapping in docker-compose.yml, as this is probably not safe:

change from 
ports:
      - "27018:27017"

to
ports:
      - "27017"
